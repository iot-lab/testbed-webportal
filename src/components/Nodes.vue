<template>
<div class="container mt-3">
  <p class="float-right mt-2 mb-4" v-if="isAdmin">
    <span class="d-block h5 text-right"><i class="fa fa-lock"></i> Admin</span>
    <a class="btn btn-warning" href="" @click.prevent="updateNodesProperties">Update nodes</a>
  </p>
  
  <h2>IoT-LAB nodes <button class="btn btn-light text-dark ml-2" @click="downloadJson" v-tooltip:right="'Download all nodes (JSON)'"><i class="fa fa-download"></i></button></h2>
  <p class="lead mb-0">Sites</p>
  <p class="mb-2" v-if="sites">
    <span class="badge badge-pill" :class="{'badge-primary': currentSite === 'all', 'badge-secondary': currentSite !== 'all'}" @click="currentSite = 'all'" style="cursor: pointer">{{sites.length}} sites</span>
    <span v-for="site in sites" class="badge badge-pill" :class="{'badge-primary': currentSite === site, 'badge-secondary': currentSite !== site}" style="margin-right: 4px; cursor: pointer"
    @click="currentSite = site">{{site.site}}</span>
  </p>
  <p class="lead mb-0">Architectures</p>
  <p class="mb-2" v-if="sites">
    <span class="badge badge-pill" :class="{'badge-primary': currentArchi === 'all', 'badge-secondary': currentArchi !== 'all'}" @click="currentArchi = 'all'" style="cursor: pointer">{{archis.length}} archis</span>
    <span v-for="archi in archis" class="badge badge-pill" :class="{'badge-primary': currentArchi === archi, 'badge-secondary': currentArchi !== archi}" style="margin-right: 4px; cursor: pointer"
    @click="currentArchi = archi">{{archi | formatArchi}}
      <span class="font-weight-normal" v-if="$options.filters.formatRadio(archi)">({{archi | formatRadio}})</span>
    </span>
  </p>
  <p class="lead mb-0">Nodes</p>
  <p class="mb-2" v-if="nodes">
    <span class="cursor badge badge-pill" :class="(nodeFilter === null) ? 'badge-primary' : 'badge-secondary'" @click="nodeFilter = null">{{getNodes().length}} nodes</span>
    <span class="cursor badge badge-pill" :class="(nodeFilter === nodeFilters.alive) ? 'badge-primary' : 'badge-success'" @click="nodeFilter = nodeFilters.alive" v-if="getNodes(['Alive']).length">available {{getNodes(['Alive']).length}}</span>
    <span class="cursor badge badge-pill" :class="(nodeFilter === nodeFilters.busy) ? 'badge-primary' : 'badge-warning'" @click="nodeFilter = nodeFilters.busy" v-if="getNodes(['Busy']).length">busy {{getNodes(['Busy']).length}}</span> 
    <span class="cursor badge badge-pill" :class="(nodeFilter === nodeFilters.unavailable) ? 'badge-primary' : 'badge-danger'" @click="nodeFilter = nodeFilters.unavailable" v-if="getNodes(['Absent','Suspected']).length">unavailable {{getNodes(['Absent','Suspected']).length}}</span>
    <span class="cursor badge badge-pill" :class="(nodeFilter === nodeFilters.dead) ? 'badge-primary' : 'badge-dark'" @click="nodeFilter = nodeFilters.dead" v-if="getNodes(['Dead']).length">dead {{getNodes(['Dead']).length}}</span>
    <span class="cursor badge badge-pill" :class="(nodeFilter === nodeFilters.mobile) ? 'badge-primary' : 'badge-info'" @click="nodeFilter = nodeFilters.mobile" v-if="getNodes().filter(node => node.mobile).length">mobile {{getNodes().filter(node => node.mobile).length}}</span>
  </p>
  <p v-else>
    <i class="fa fa-spinner fa-spin fa-fw"></i>
  </p>
  <div class="row">
    <div class="col-md-8">
      <label class="mt-2 text-muted font-italic">Showing {{filteredNodes.length}} nodes</label>
    </div>
    <div class="col-md-4">
      <input type="text" class="form-control mb-3" placeholder="Search hostname or uid" v-model="search">
    </div>
  </div>
  <table class="table table-striped table-sm" v-if="nodes.length">
    <thead>
      <tr>
        <th class="cursor" title="sort by hostname" @click="sortBy(node => nodeSortByHostname(node))">Node hostname</th>
        <th class="cursor text-center" title="sort by state" @click="sortBy(node => node.state)">State</th>
        <th class="cursor" title="sort by archi" @click="sortBy(node => node.archi)">Archi <span class="text-muted font-weight-normal">(radio)</span></th>
        <th class="cursor" title="sort by site" @click="sortBy(node => node.site)">Site</th>
        <th class="cursor" title="sort by mobility" @click="sortBy(node => node.mobility_type, reverse = true)">Mobility</th>
        <th class="cursor" title="sort by uid" @click="sortBy(node => node.uid)">UID</th>
        <th class="cursor" title="sort by X" @click="sortBy(node => node.x, reverse = true)">X</th>
        <th class="cursor" title="sort by Y" @click="sortBy(node => node.y, reverse = true)">Y</th>
        <th class="cursor" title="sort by Z" @click="sortBy(node => node.z, reverse = true)">Z</th>
      </tr>
    </thead>
    <tbody>
      <tr v-for="node in filteredNodes">
        <td>{{node.network_address}}</td>
        <td class="text-center"><span class="badge badge-state" :class="node.state | stateBadgeClass">{{node.state}}</span></td>
        <td class="text-capitalize-first">{{node.archi | formatArchi}}
          <span class="text-muted" v-if="$options.filters.formatRadio(node.archi)">({{node.archi | formatRadio}})</span>
        </td>
        <td class="text-capitalize">{{node.site}}</td>
        <td v-if="node.mobile">Yes <span class="text-muted">({{node.mobility_type}})</span></td>
        <td v-else></td>
        <td>{{node.uid}}</td>
        <td>{{node.x}}</td>
        <td>{{node.y}}</td>
        <td>{{node.z}}</td>
      </tr>
    </tbody>
  </table>

</div> <!-- container -->

</template>

<script>
import { iotlab } from '@/rest'
import { auth } from '@/auth'
import { downloadObjectAsJson } from '@/utils'
import json2csv from 'json2csv'

export default {
  name: 'Nodes',

  data () {
    return {
      isAdmin: auth.isAdmin,
      sites: [],
      nodes: [],
      currentSite: 'all',
      currentArchi: 'all',
      nodeFilter: null,
      nodeFilters: {
        alive: node => node.state === 'Alive',
        busy: node => node.state === 'Busy',
        unavailable: node => ['Absent', 'Suspected'].includes(node.state),
        dead: node => node.state === 'Dead',
        mobile: node => node.mobile,
      },
      search: '',
    }
  },

  created () {
    iotlab.getSitesDetails().then(data => { this.sites = data.sort((a, b) => a.site.localeCompare(b.site)) })
    iotlab.getNodes().then(data => { this.nodes = data })
  },

  computed: {
    archis () {
      let archis
      if (this.currentSite === 'all') {
        archis = Array.from(this.sites.reduce((acc, site) => {
          site.archis.map(archi => acc.add(archi.archi))
          return acc
        }, new Set()))
      } else {
        archis = this.sites.find(site => site.site === this.currentSite.site).archis.map(archi => archi.archi)
      }
      return archis.sort((a, b) => a.localeCompare(b))
    },
    filteredNodes () {
      let nodes = this.getNodes()
      if (this.search) nodes = nodes.filter(node => node.network_address.includes(this.search) || node.uid.includes(this.search))
      if (this.nodeFilter) nodes = nodes.filter(this.nodeFilter)
      if (this.currentSite !== 'all') nodes = nodes.filter(node => node.site === this.currentSite.site)
      if (this.currentArchi !== 'all') nodes = nodes.filter(node => node.archi === this.currentArchi)
      return nodes
    },
  },

  methods: {
    async downloadJson () {
      console.log(json2csv.parse(await iotlab.getNodes()))
      downloadObjectAsJson(await iotlab.getNodes(), 'iotlab-nodes')
    },

    getNodes (stateList = null) {
      let nodes = this.nodes
      if (this.currentSite !== 'all') nodes = nodes.filter(node => node.site === this.currentSite.site)
      if (this.currentArchi !== 'all') nodes = nodes.filter(node => node.archi === this.currentArchi)
      if (stateList) {
        return nodes.filter(node => stateList.includes(node.state))
      } else {
        return nodes
      }
    },

    sortBy (func, reverse = false) {
      this.nodes = this.nodes.sort((a, b) => {
        if (func(a) === func(b)) return this.nodeSortByHostname(a).localeCompare(this.nodeSortByHostname(b))
        if (typeof func(a) === 'string' || typeof func(b) === 'string') return func(a).localeCompare(func(b))
        if (typeof func(a) === 'number' || typeof func(b) === 'number') return func(a) - func(b)
        console.log('Unhandled sort criteria', a.network_address, func(a), b.network_address, func(b))
      })
      if (reverse) this.nodes.reverse()
    },

    // Allow sorting of hostnames by domain then by servername
    // the trick is to append the servername after the domain (e.g 'a8-1.grenoble.iot-lab.info' -> 'grenoble.iot-lab.info.a8-1')
    // then it's possible to simply sort the strings
    nodeSortByHostname (node) {
      let chuncks = node.network_address.split('.')
      chuncks.push(chuncks.shift()) // or we could use chunks.reverse()
      return chuncks.join('.')
    },

    async updateNodesProperties () {
      if (!confirm('This can take a while. Are you sure?')) return
      try {
        await iotlab.updateNodesProperties()
        this.$notify({text: 'Nodes properties updated', type: 'success'})
      } catch (err) {
        this.$notify({text: 'An error occured', type: 'error'})
      }
    },

  },

}
</script>

<style scoped>
.text-capitalize-first:first-letter {
  text-transform: uppercase;
}
.badge {
  font-size: 80%;
}
</style>
