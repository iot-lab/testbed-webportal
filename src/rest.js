import axios from 'axios'

const API_URL = `https://${process.env.VUE_APP_IOTLAB_HOST}/api/`

export const iotlab = {

  api: axios.create({baseURL: API_URL, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')}),

  create (username, password) {
    this.api = axios.create({
      baseURL: API_URL,
      auth: {
        username: username,
        password: password,
      },
    })
  },

  // USERS API

  async signup (user, nbUsers) {
    if (nbUsers) {
      // create multiple user accounts
      await iotlab.api.post(`/users?nbusers=${nbUsers}`, user)
    } else {
      // create single user account
      await iotlab.api.post('/users', user)
    }
  },

  async activateAccount (hash) {
    await iotlab.api.post('/users/activate/', {'hash': hash})
  },

  async deleteUser (login = undefined) {
    if (login) {
      await iotlab.api.delete(`/users/${login}`)
    } else {
      await iotlab.api.delete(`/user`)
    }
  },

  async resetPassword (email) {
    await iotlab.api.post(`/users/reset_password`, {'email': email})
  },

  async changePassword (oldPwd, newPwd, confirmPwd) {
    await iotlab.api.put('/user/password', {
      'old_password': oldPwd,
      'new_password': newPwd,
      'confirm_new_password': confirmPwd,
    })
  },

  async getSSHkeys () {
    return iotlab.api.get('/user/keys').then(resp => resp.data.sshkeys)
  },

  async modifySSHkeys (keys) {
    return iotlab.api.post('/user/keys', {'sshkeys': keys})
  },

  async deleteSSHkey (id) {
    return iotlab.api.delete(`/user/keys/${id}`)
  },

  async getUsers ({status, group, search} = {}) {
    let params = {}
    if (status) {
      params.status = status
    }
    if (group) {
      params.group = group
    }
    if (search) {
      params.search = search
    }
    return iotlab.api.get('/users', {params: params}).then(resp => resp.data)
  },

  async getUserInfo (login = undefined) {
    if (login) {
      return iotlab.api.get(`/users/${login}`).then(resp => resp.data)
    } else {
      return iotlab.api.get('/user').then(resp => resp.data)
    }
  },

  async setUserInfo (user, login = undefined) {
    if (login) {
      return iotlab.api.put(`/users/${login}`, user)
    } else {
      return iotlab.api.put('/user', user)
    }
  },

  async activateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'active'})
  },

  async deactivateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'inactive'})
  },

  async setUserGroups (login, groups) {
    return iotlab.api.put(`/users/${login}`, {'groups': groups})
  },

  async sendMail (recipients, subject, message) {
    await iotlab.api.post('/users/email', {
      recipients: recipients,
      subject: subject,
      body: message,
    })
  },

  // EXPERIMENT API

  async getUserExperimentsCount () {
    return iotlab.api.get('/experiments/total').then(resp => resp.data)
  },

  async getAllExperimentsCount (user) {
    if (user) {
      return iotlab.api.get(`/experiments/total/users/${user}`).then(resp => resp.data)
    } else {
      return iotlab.api.get(`/experiments/total/all`).then(resp => resp.data)
    }
  },

  async getAllExperiments ({limit, offset, user, state = 'Terminated,Error,Running,Finishing,Resuming,toError,Waiting,Launching,Hold,toLaunch,toAckReservation,Suspended'} = {}) {
    let params = {}
    if (state) {
      params.state = state
    }
    if (limit) {
      params.limit = Math.min(limit, 500) // IoT-LAB api restrict queries to 500 entries
    }
    if (offset) {
      params.offset = offset
    }
    if (user === '@self') {
      return iotlab.api.get('/experiments', {params: params}).then(resp => resp.data.items)
    } else if (user) {
      return iotlab.api.get(`/experiments/users/${user}`, {params: params}).then(resp => resp.data.items)
    } else {
      return iotlab.api.get('/experiments/all', {params: params}).then(resp => resp.data.items)
    }
  },

  async getRunningExperiments () {
    return iotlab.api.get('/experiments/running').then(resp => resp.data.items)
  },

  async getExperiment (id) {
    return iotlab.api.get(`/experiments/${id}`).then(resp => resp.data)
  },

  async getExperimentArchive (id) {
    return iotlab.api.get(`/experiments/${id}/data`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async getExperimentDeployment (id) {
    return iotlab.api.get(`/experiments/${id}/deployment`).then(resp => resp.data)
  },

  async stopExperiment (id) {
    return iotlab.api.delete(`/experiments/${id}`)
  },

  async submitExperiment ({type, name, duration, reservation, nodes, firmwareassociations, firmwares, profileassociations, profiles} = {}) {
    const formData = new FormData()
    formData.append('experiment.json', JSON.stringify({
      type: type,
      duration: duration,
      reservation: reservation,
      name: name,
      nodes: nodes,
      firmwareassociations: firmwareassociations,
      profileassociations: profileassociations,
    }))
    // attached firmware names should be unique
    let attachedFirmwares = []
    for (let asso of firmwareassociations || []) {
      if (asso.firmwarename in firmwares && !attachedFirmwares.includes(asso.firmwarename)) {
        formData.append(asso.firmwarename, new Blob([firmwares[asso.firmwarename]], {type: 'application/octet-stream'}), asso.firmwarename)
        attachedFirmwares.push(asso.firmwarename)
      }
    }
    return iotlab.api.post('/experiments', formData).then(resp => resp.data)
  },

  async checkFirmware (fileString) {
    const formData = new FormData()
    formData.append('firmware.bin', fileString)
    return iotlab.api.post('/firmwares/checker', formData).then(resp => resp.data)
  },

  async flashFirmware (id, nodes, fileString) {
    const formData = new FormData()
    formData.append(`experiment${id}.json`, JSON.stringify(nodes))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/experiments/${id}/nodes/flash`, formData).then(resp => resp.data)
  },

  async flashResourcesFirmware (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/flash/${name}`, nodes).then(resp => resp.data)
  },

  async updateMonitoring (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/monitoring/${name}`, nodes).then(resp => resp.data)
  },

  async updateExperimentMobilityCircuit (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/robots/mobility/${name}`, nodes).then(resp => resp.data)
  },

  async reloadExperiment (id) {
    return iotlab.api.post(`/experiments/${id}/reload`, {}).then(resp => resp.data)
  },

  // USER GROUPS

  async getUserGroups () {
    return iotlab.api.get('/groups').then(resp => resp.data)
  },

  async createGroup (name) {
    return iotlab.api.post(`/groups`, {name: name})
  },

  async deleteGroup (group) {
    return iotlab.api.delete(`/groups/${group}`)
  },

  // OTHER API

  async getSites () {
    return iotlab.api.get('/sites').then(resp => resp.data.items)
  },

  async getSitesDetails () {
    return iotlab.api.get('/sites/details').then(resp => resp.data.items)
  },

  async getNodes () {
    return iotlab.api.get('/nodes').then(resp => resp.data.items)
  },

  async updateNodesProperties () {
    return iotlab.api.put('/nodes')
  },

  async updateNodesPropertiesOldApi () {
    const OLD_API = `https://${process.env.VUE_APP_IOTLAB_HOST}/rest/`
    const oldApi = axios.create({baseURL: OLD_API, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')})
    return oldApi.post('/admin/nodes')
  },

  async getNodesIds () {
    return iotlab.api.get('/nodes/ids').then(resp => resp.data.items)
  },

  async getExperimentNodes (id) {
    return iotlab.api.get(`/experiments/${id}/nodes`).then(resp => resp.data.items)
  },

  async getExperimentToken (id) {
    return iotlab.api.get(`/experiments/${id}/token`).then(resp => resp.data.token)
  },

  async sendNodesCommand (id, cmd, nodes) {
    // cmd in [start, stop, reset, flash-idle, profile-reset, debug-start, debug-stop]
    return iotlab.api.post(`/experiments/${id}/nodes/${cmd}`, nodes).then(resp => resp.data)
  },

  // STORES (firmwares, profiles, etc)

  async getMobilityCircuits () {
    return iotlab.api.get('/mobilities/circuits').then(resp => resp.data.items)
  },

  async getMobilityCircuit (name) {
    return iotlab.api.get(`/mobilities/circuits/${name}`).then(resp => resp.data)
  },

  async getRobotsSiteMapImage (site) {
    return iotlab.api.get(`/robots/${site}/map/image`, {
      responseType: 'arraybuffer',
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    }).then(resp => Buffer.from(resp.data, 'binary').toString('base64')
    )
  },

  async getRobotsSiteMapConfig (site) {
    return iotlab.api.get(`/robots/${site}/map/config`).then(resp => resp.data)
  },

  async createMobilityCircuit (circuit) {
    return iotlab.api.post(`/mobilities/circuits`, circuit)
  },

  async updateMobilityCircuit (name, circuit) {
    return iotlab.api.put(`/mobilities/circuits/${name}`, circuit)
  },

  async deleteMobilityCircuit (name) {
    return iotlab.api.delete(`/mobilities/circuits/${name}`)
  },

  async getMonitoringProfiles () {
    return iotlab.api.get('/monitoring').then(resp => resp.data)
  },

  async getMonitoringProfile (name) {
    return iotlab.api.get(`/monitoring/${name}`).then(resp => resp.data)
  },

  async updateMonitoringProfile (name, profile) {
    return iotlab.api.put(`/monitoring/${name}`, profile)
  },

  async createMonitoringProfile (profile) {
    return iotlab.api.post(`/monitoring`, profile)
  },

  async deleteMonitoringProfile (name) {
    return iotlab.api.delete(`/monitoring/${name}`)
  },

  async getFirmwares () {
    return iotlab.api.get('/firmwares').then(resp => resp.data)
  },

  async getFirmware (name) {
    return iotlab.api.get(`/firmwares/${name}`).then(resp => resp.data)
  },

  async getFirmwareFile (name) {
    return iotlab.api.get(`/firmwares/${name}/file`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async updateFirmware (name, firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    if (fileString) {
      formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    }
    return iotlab.api.put(`/firmwares/${name}`, formData).then(resp => resp.data)
  },

  async createFirmware (firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/firmwares`, formData).then(resp => resp.data)
  },

  async deleteFirmware (name) {
    return iotlab.api.delete(`/firmwares/${name}`)
  },
}
