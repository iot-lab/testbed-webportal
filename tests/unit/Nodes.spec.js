import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import Nodes from '@/views/Nodes'
import { iotlab } from '@/rest'
jest.mock('@/rest')

const localVue = createLocalVue()
localVue.use(VueRouter)
const router = new VueRouter()

describe('Nodes.vue', () => {

  it('should match snapshot', () => {
    const wrapper = shallowMount(Nodes, {localVue, router})
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

})
