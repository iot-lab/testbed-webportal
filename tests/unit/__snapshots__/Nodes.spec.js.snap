// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nodes.vue should match snapshot 1`] = `
<div
  class="container mt-3"
>
  <div
    class="float-right mt-1 mb-4"
  >
    <div
      class="dropdown d-inline-block "
    >
      <button
        aria-expanded="false"
        aria-haspopup="true"
        class="btn btn-light mr-1"
        data-toggle="dropdown"
      >
        <i
          class="fa fa-fw fa-download"
        />
         Download
      </button>
       
      <div
        aria-labelledby="dropdownMenuButton"
        class="dropdown-menu"
      >
        <a
          class="dropdown-item"
          href="#"
        >
          IoT-LAB nodes 
          <span
            class="badge badge-pill badge-info"
          >
            JSON
          </span>
        </a>
         
        <a
          class="dropdown-item"
          href="#"
        >
          IoT-LAB nodes 
          <span
            class="badge badge-pill badge-info"
          >
            CSV
          </span>
        </a>
      </div>
    </div>
     
    <!---->
  </div>
   
  <h2>
    IoT-LAB nodes
  </h2>
   
  <p
    class="lead mb-0"
  >
    Sites
  </p>
   
  <p
    class="mb-2"
  >
    <span
      class="badge badge-pill mr-1 cursor badge-primary"
    >
      0 sites
    </span>
     
  </p>
   
  <p
    class="lead mb-0"
  >
    Architectures
  </p>
   
  <p
    class="mb-2"
  >
    <span
      class="badge badge-pill mr-1 cursor badge-primary"
    >
      0 archis
    </span>
     
  </p>
   
  <p
    class="lead mb-0"
  >
    Nodes
  </p>
   
  <p
    class="mb-2"
  >
    <span
      class="badge badge-pill mr-1 cursor badge-primary"
    >
      0 nodes
    </span>
     
    <!---->
     
    <!---->
     
    <!---->
     
    <!---->
     
    <!---->
  </p>
   
  <div
    class="row"
  >
    <div
      class="col-md-8"
    >
      <label
        class="mt-2 text-muted font-italic"
      >
        Showing 0 nodes
      </label>
    </div>
     
    <div
      class="col-md-4"
    >
      <input
        class="form-control mb-3"
        placeholder="Search hostname or uid"
        type="text"
      />
    </div>
  </div>
   
  <map3d-stub />
   
  <!---->
</div>
`;
